### PR TITLE
scheduler CC: add v1beta2 API, deprecate plugins

### PR DESCRIPTION
**Pull request by adtac**
#### What type of PR is this?

/kind api-change

#### What this PR does / why we need it:

Adds scheduler config v1beta2 API. This PR builds on top of https://github.com/kubernetes/kubernetes/pull/95308.

xref: https://github.com/kubernetes/kubernetes/issues/95446 - things completed in this PR:

* [x]  Deprecate some Plugins #94008
* [ ]  Deprecate resources plugins in favor of a unified Fit plugin #98746 (not doing this in this PR)
* [x]  Remove backwards compatibility hacks:
  * [x]  Always set DefaultingType to System (PodTopologySpread). Hack added in #95048
  * [x]  Stop swallowing invalid healthz and metrics address #97048
* [ ]  [Nice-to-have] Make Plugin names fully qualified #95445 (not doing this in this PR)

#### Which issue(s) this PR fixes:

Related to https://github.com/kubernetes/kubernetes/issues/94008

#### Special notes for your reviewer:

Commits in @hprateek43's PR are preserved as-is. See everything after the 3rd commit for the changes I've made on top.

#### Does this PR introduce a user-facing change?

```release-note
kube-scheduler component config v1beta2 API available
Three scheduler plugins deprecated (NodeLabel, ServiceAffinity, NodePreferAvoidPods)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

cc @hprateek43 @alculquicondor @ahg-g @liggitt

### CODE REVIEW

File: cmd/kube-scheduler/app/options/options_test.go [hunk 1]
```diff
@@ -373,7 +443,8 @@ profiles:
 					AllowCacheTTL:                10 * time.Second,
 					DenyCacheTTL:                 10 * time.Second,
 					RemoteKubeConfigFileOptional: true,
-					AlwaysAllowPaths:             []string{"/healthz"}, // note: this does not match /healthz/ or /healthz/*
+					AlwaysAllowPaths:             []string{"/healthz", "/readyz", "/livez"}, // note: this does not match /healthz/ or /healthz/*
+					AlwaysAllowGroups:            []string{"system:masters"},
```
Comment (adtac) [inline cmd/kube-scheduler/app/options/options_test.go:447]: @hprateek43 I changed this back to make sure v1beta1 and v1beta2 tests were similar. Was there a reason for your change that I missed?

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 2]
```diff
@@ -55,14 +53,14 @@ func SetDefaults_KubeSchedulerConfiguration(obj *v1beta2.KubeSchedulerConfigurat
 	// Only apply a default scheduler name when there is a single profile.
 	// Validation will ensure that every profile has a non-empty unique name.
 	if len(obj.Profiles) == 1 && obj.Profiles[0].SchedulerName == nil {
-		obj.Profiles[0].SchedulerName = pointer.StringPtr(api.DefaultSchedulerName)
```
Comment (adtac) [inline pkg/scheduler/apis/config/v1beta2/defaults.go]: @hprateek43 did you mean to leave it as v1.DefaultSchedulerName? (k8s.io/kubernetes/pkg/apis/core.DefaultSchedulerName) is undefined.

**Review by adtac — COMMENTED** [review]
Just to double-check, I did a diff of all files in v1beta1 and v1beta2 and here's a list of all differences minus v1beta1/v1beta2 differences: http://ix.io/2RjG/diff

File: pkg/scheduler/apis/config/v1beta2/defaults.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta2/defaults.go]: that was removed recently

File: cmd/kube-scheduler/app/options/options_test.go [hunk 3]
```diff
@@ -34,11 +34,26 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	apiserveroptions "k8s.io/apiserver/pkg/server/options"
 	componentbaseconfig "k8s.io/component-base/config"
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
 	"k8s.io/component-base/logs"
+	configv1beta1 "k8s.io/kube-scheduler/config/v1beta1"
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go:39]: nit: can we have these without aliases?
Or at least the aliases don't have to say "kubescheduler", given the location of the file, if there are no conflicts.

File: cmd/kube-scheduler/app/options/options_test.go [hunk 4]
```diff
@@ -272,7 +342,7 @@ profiles:
 		checkErrFn       func(err error) bool
 	}{
 		{
-			name: "config file",
+			name: "config file", // v1beta2
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go:345]: any reason not to make v1beta2 part of the test name?

File: pkg/scheduler/apis/config/v1beta2/conversion.go [hunk 5]
```diff
@@ -46,6 +46,98 @@ func getPluginArgConversionScheme() *runtime.Scheme {
 	return pluginArgConversionScheme
 }
 
+func Convert_v1beta2_Plugins_To_config_Plugins(in *v1beta2.Plugins, out *config.Plugins, s conversion.Scope) error {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta2/conversion.go:49]: What if we get rid of pointers?

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 6]
```diff
@@ -80,8 +78,9 @@ func SetDefaults_KubeSchedulerConfiguration(obj *v1beta2.KubeSchedulerConfigurat
 				hostPort := net.JoinHostPort(*obj.HealthzBindAddress, strconv.Itoa(config.DefaultInsecureSchedulerPort))
 				obj.HealthzBindAddress = &hostPort
 			} else {
-				// TODO: in v1beta2 we should let this error instead of stomping with a default value
-				obj.HealthzBindAddress = &defaultBindAddress
+				// This is a malformed partial address and there's nothing we can do
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta2/defaults.go:81]: Isn't there any guards in validation.go?
Also I don't think the comment is necessary.

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 7]
```diff
@@ -211,7 +211,11 @@ func SetDefaults_VolumeBindingArgs(obj *v1beta2.VolumeBindingArgs) {
 func SetDefaults_PodTopologySpreadArgs(obj *v1beta2.PodTopologySpreadArgs) {
 	if feature.DefaultFeatureGate.Enabled(features.DefaultPodTopologySpread) {
 		if obj.DefaultingType == "" {
-			obj.DefaultingType = v1beta2.SystemDefaulting
+			if len(obj.DefaultConstraints) != 0 {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/v1beta2/defaults.go:214]: why revert this? It was intended. Defaults should be independent of other fields.

**Review by alculquicondor — COMMENTED** [review]
Looking good. Could you start a PR for #94008 already?

File: pkg/scheduler/apis/config/v1beta2/conversion.go (hunk 5 above)
Comment (adtac) [reply pkg/scheduler/apis/config/v1beta2/conversion.go:49]: do you mean `type Plugins struct { QueueSort PluginSet }`? instead of `*PluginSet` in the v1beta2 API?

File: pkg/scheduler/apis/config/v1beta2/defaults.go (hunk 7 above)
Comment (adtac) [reply pkg/scheduler/apis/config/v1beta2/defaults.go:214]: Oops, good catch, I missed this when reviewing.

File: pkg/scheduler/apis/config/v1beta2/defaults.go (hunk 6 above)
Comment (adtac) [reply pkg/scheduler/apis/config/v1beta2/defaults.go:81]: You're right, it is caught by validation. However we didn't test this scenario in validation_test.go, so I've added a test for this.

File: cmd/kube-scheduler/app/options/options_test.go (hunk 4 above)
Comment (alculquicondor) [reply cmd/kube-scheduler/app/options/options_test.go:345]: better to have more context when I'm running the test.

File: pkg/scheduler/apis/config/v1beta2/conversion.go (hunk 5 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta2/conversion.go:49]: correct

File: pkg/scheduler/apis/config/v1beta2/defaults.go (hunk 6 above)
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta2/defaults.go]: awesome

File: cmd/kube-scheduler/app/options/options_test.go [hunk 8]
```diff
@@ -34,21 +34,21 @@ import (
...
 	apiserveroptions "k8s.io/apiserver/pkg/server/options"
 	componentbaseconfig "k8s.io/component-base/config"
-	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+	"k8s.io/component-base/config/v1alpha1"
 	"k8s.io/component-base/logs"
-	configv1beta1 "k8s.io/kube-scheduler/config/v1beta1"
+	"k8s.io/kube-scheduler/config/v1beta1"
 	kubeschedulerconfig "k8s.io/kubernetes/pkg/scheduler/apis/config"
-	kubeschedulerscheme "k8s.io/kubernetes/pkg/scheduler/apis/config/scheme"
+	scheme "k8s.io/kubernetes/pkg/scheduler/apis/config/scheme"
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/options/options_test.go:41]: remove the alias when it matches the package name.

**Comment by adtac** [comment]
I've squashed some commits and added a new commit. The second commit in this PR adds the infrastructure to deprecate plugins in CC; in v1beta2, we want to deprecate NodeLabel and ServiceAffinity. The second commit was a bit tricky from a API point-of-view, so please lmk if I'm doing something wrong.

**Comment by liggitt** [comment]
> I know we're past the 1.21 code freeze, so we'll be targetting v1beta2 in 1.22 (with possibly more changes than there are in this PR)

sounds good

/milestone v1.22

File: pkg/scheduler/apis/config/types_pluginargs.go [hunk 9]
```diff
@@ -56,7 +56,9 @@ type InterPodAffinityArgs struct {
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// NodeLabelArgs holds arguments used to configure the NodeLabel plugin.
+// NodeLabelArgs holds arguments used to configure the NodeLabel plugin. Note:
+// this plugin has been deprecated and is only configurable through the
+// scheduler policy API.
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types_pluginargs.go:61]: but also v1beta1, isn't it?

File: pkg/scheduler/apis/config/validation/validation.go [hunk 10]
```diff
@@ -80,6 +83,80 @@ func ValidateKubeSchedulerConfiguration(cc *config.KubeSchedulerConfiguration) f
...
+		if removed, removedVersion := isPluginRemoved(apiVersion, plugin.Name); removed {
+			allErrs = append(allErrs, field.Invalid(path.Child("enabled").Index(i), plugin.Name, fmt.Sprintf("was removed in version %q (KubeSchedulerConfiguration is version %q)", removedVersion, apiVersion)))
+		}
+	}
+	return allErrs
+}
+
+func validateProfilesForRemovedPlugins(path *field.Path, apiVersion string, profiles []config.KubeSchedulerProfile) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for i, profile := range profiles {
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/validation/validation.go:129]: shouldn't this be part of `validateKubeSchedulerProfile` instead?

File: pkg/scheduler/framework/runtime/framework.go [hunk 11]
```diff
@@ -402,13 +412,21 @@ func registerClusterEvents(name string, eventToPlugins map[framework.ClusterEven
...
 	if ok {
 		return res, nil
 	}
 	// Use defaults from latest config API version.
-	gvk := v1beta2.SchemeGroupVersion.WithKind(name + "Args")
+	var gvk schema.GroupVersionKind
+	switch componentConfigVersion {
+	case v1beta1.SchemeGroupVersion.String():
+		gvk = v1beta1.SchemeGroupVersion.WithKind(name + "Args")
+	case v1beta2.SchemeGroupVersion.String():
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:425]: we should do empty equivalent to v1beta2, otherwise we have to specify the version in every test.

File: pkg/scheduler/apis/config/validation/validation.go (hunk 10 above)
Comment (adtac) [reply pkg/scheduler/apis/config/validation/validation.go]: Good point, moved it inside that function

File: pkg/scheduler/apis/config/v1beta2/conversion.go (hunk 5 above)
Comment (adtac) [reply pkg/scheduler/apis/config/v1beta2/conversion.go:49]: Done. (See fourth commit.)

**Comment by adtac** [comment]
@liggitt this is ready for API review

File: pkg/scheduler/apis/config/types_pluginargs.go [hunk 12]
```diff
@@ -180,7 +182,9 @@ type ResourceSpec struct {
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// ServiceAffinityArgs holds arguments used to configure the ServiceAffinity plugin.
+// ServiceAffinityArgs holds arguments used to configure the ServiceAffinity
+// plugin. Note: this plugin has been deprecated and is only configurable
```
Comment (alculquicondor) [inline pkg/scheduler/apis/config/types_pluginargs.go:186]: nit: https://blog.golang.org/godoc#:~:text=Deprecated

**Review by alculquicondor — COMMENTED** [review]
It looks like you missed one plugin from the release note.

Can you list in the PR description what things are being done here from #95446?

File: cmd/kube-scheduler/app/options/configfile.go [hunk 13]
```diff
@@ -37,32 +37,32 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
...
 
 	const mediaType = runtime.ContentTypeYAML
-	info, ok := runtime.SerializerInfoForMediaType(kubeschedulerscheme.Codecs.SupportedMediaTypes(), mediaType)
+	info, ok := runtime.SerializerInfoForMediaType(scheme.Codecs.SupportedMediaTypes(), mediaType)
 	if !ok {
 		return fmt.Errorf("unable to locate encoder -- %q is not a supported media type", mediaType)
 	}
 
-	encoder := kubeschedulerscheme.Codecs.EncoderForVersion(info.Serializer, kubeschedulerconfigv1beta1.SchemeGroupVersion)
+	encoder := scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta2.SchemeGroupVersion)
```
[the remark this answers is not in the document]
Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go]: Done. I've also modified the tests to test this scenario.

**Review by alculquicondor — COMMENTED** [review]
Include all the relevant changes in the release note.

File: cmd/kube-scheduler/app/options/configfile.go [hunk 14]
```diff
@@ -37,39 +40,55 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
...
+	var encoder runtime.Encoder
+	switch cfg.ComponentConfigVersion {
+	case configv1beta1.SchemeGroupVersion.String():
+		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta1.SchemeGroupVersion)
+	case configv1beta2.SchemeGroupVersion.String():
+		encoder = scheme.Codecs.EncoderForVersion(info.Serializer, configv1beta2.SchemeGroupVersion)
+	default:
+		return buf, fmt.Errorf("unexpected component config version %q", cfg.ComponentConfigVersion)
+	}
+	if err := encoder.Encode(cfg, buf); err != nil {
```
[the remark this answers is not in the document]
Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go:72]: runtime.Encode (returns []byte) uses encoder.Encode (returns bytes.Buffer) internally; it's just a small wrapper. I just liked the bytes.Buffer interface better than dealing with raw []byte

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+	if obj.HealthzBindAddress == nil {
+		obj.HealthzBindAddress = &defaultBindAddress
+	} else {
+		if host, port, err := net.SplitHostPort(*obj.HealthzBindAddress); err == nil {
+			if len(host) == 0 {
+				host = "0.0.0.0"
+			}
+			hostPort := net.JoinHostPort(host, port)
+			obj.HealthzBindAddress = &hostPort
+		} else {
```
[the remark this answers is not in the document]
Comment (adtac) [reply pkg/scheduler/apis/config/v1beta2/defaults.go:74]: No, what we wanted to stop doing was this:

https://github.com/kubernetes/kubernetes/blob/5b46064cae41395cf775cb445ffdaae1261076c7/pkg/scheduler/apis/config/v1beta1/defaults.go#L79-L82

The else clause in this PR is there so that plain IPs are allowed in addition to IP:port. We no longer allow inputs that are invalid IP and invalid IP:port, which is what that issue is talking about.

I'll update the release note and mark the completed items in a bit.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/apis/config/v1beta2/defaults.go:74]: ahh, it was the next else. Sorry :)

**Comment by adtac** [comment]
@alculquicondor I've updated the PR description with a checklist of things done in this PR. Also, I just realised I had already deprecated NodePreferAvoidPods in this PR but I forgot about it :sweat_smile: (context: we discussed this offline recently)

File: pkg/scheduler/framework/plugins/nodelabel/node_label.go [hunk 16]
```diff
@@ -14,6 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// Deprecated: this plugin has been deprecated and is only configurable through
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/nodelabel/node_label.go:17]: maybe we can also add a log line warning about the use of a deprecated plugin in v1beta1.

Comment (adtac) [reply pkg/scheduler/framework/plugins/nodelabel/node_label.go:17]: ack. I've kept the message simple.

**Comment by alculquicondor** [comment]
/approve

please squash

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+	// Validation will ensure that every profile has a non-empty unique name.
+	if len(obj.Profiles) == 1 && obj.Profiles[0].SchedulerName == nil {
+		obj.Profiles[0].SchedulerName = pointer.StringPtr(v1.DefaultSchedulerName)
+	}
+
+	// For Healthz and Metrics bind addresses, we want to check:
+	// 1. If the value is nil, default to 0.0.0.0 and default scheduler port
+	// 2. If there is a value set, attempt to split it. If it's just a port (ie, ":1234"), default to 0.0.0.0 with that port
+	// 3. If splitting the value fails, check if the value is even a valid IP. If so, use that with the default port.
+	// Otherwise leave the address as-is, it will be caught during validation.
```
[the remark this answers is not in the document]
Comment (adtac) [reply pkg/scheduler/apis/config/v1beta2/defaults.go:63]: good catch, added a test (see "metrics and healthz with invalid IP" in defaults_test.go)

File: pkg/scheduler/apis/config/validation/validation.go [hunk 18]
```diff
@@ -80,16 +82,57 @@ func ValidateKubeSchedulerConfiguration(cc *config.KubeSchedulerConfiguration) u
...
+		schemeGroupVersion: v1beta1.SchemeGroupVersion.String(),
+		plugins:            []string{},
+	},
+	{
+		schemeGroupVersion: v1beta2.SchemeGroupVersion.String(),
+		plugins:            []string{"NodeLabel", "ServiceAffinity", "NodePreferAvoidPods"},
+	},
+}
+
+func isPluginRemoved(componentConfigVersion string, name string) (bool, string) {
```
[the remark this answers is not in the document]
Comment (adtac) [reply pkg/scheduler/apis/config/validation/validation.go:101]: >behavior for future versions

could you elaborate?

File: pkg/scheduler/apis/config/types.go [hunk 19]
```diff
@@ -49,6 +49,10 @@ const (
 type KubeSchedulerConfiguration struct {
 	metav1.TypeMeta
 
+	// ComponentConfigVersion is the external type version that should be used to
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/scheduler/apis/config/types.go:52]: *if* you're going to set a field in conversion (which is unusual), is there a reason not to set the TypeMeta.APIVersion field which already exists to hold this info?

File: pkg/scheduler/apis/config/types_pluginargs.go (hunk 12 above)
Comment (adtac) [reply pkg/scheduler/apis/config/types_pluginargs.go]: so I reverted this change because of staticcheck, which checks for uses of symbols that have "Deprecated:" in their docs. Since we still use these types in a lot of places in the scheduler code, I think it makes sense to back to not using "Deprecated:". WDYT?

File: pkg/scheduler/apis/config/types.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (adtac) [reply pkg/scheduler/apis/config/types.go:52]: So I did consider TypeMeta, but I was honestly expecting it to be filled by API machinery - is that not the case? (TypeMeta.APIVersion was empty)

If TypeMeta.APIVersion is not apimachinery-reserved or anything like that, I agree with you

[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/scheduler/apis/config/types.go:52]: When converting from external → internal, the internal object is not expected to bear any trace of the originating external version, so APIVersion is cleared.

This is explicitly going against that convention by adding a new internal field specifically to track the originating external version. If it's going to do that, I can't think of a reason not to use the field already present to hold APIVersion info

**Comment by ahg-g** [comment]
Are we close to merge this PR? do you need help moving it forward?

File: pkg/scheduler/apis/config/validation/validation.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/scheduler/apis/config/validation/validation.go:106]: I think the comment on `removedPlugins` addresses this thanks!

File: pkg/scheduler/apis/config/types_pluginargs.go [hunk 20]
```diff
@@ -57,6 +57,10 @@ type InterPodAffinityArgs struct {
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
 // NodeLabelArgs holds arguments used to configure the NodeLabel plugin.
+//
+// This plugin has been deprecated and is only configurable through the
+// scheduler policy API and the v1beta1 component config API. It is recommended
+// to use the NodeAffinity plugin instead.
```
Comment (msau42) [inline pkg/scheduler/apis/config/types_pluginargs.go:63]: Can we also add this deprecation notice on the v1beta1 api?

**Review by msau42 — COMMENTED** [review]
From an API standpoint, the main thing that is remaining is to resolve https://github.com/kubernetes/kubernetes/pull/99597/files#r631975169

**Comment by ahg-g** [comment]
> From an API standpoint, the main thing that is remaining is to resolve https://github.com/kubernetes/kubernetes/pull/99597/files#r631975169

It seems Adi is having some issues resolving this. Can we address it in a followup PR? This is not a user facing change, so it is safe to address later. I would like to get this PR merged asap because it is blocking multiple followup work.

**Comment by ahg-g** [comment]
> > From an API standpoint, the main thing that is remaining is to resolve https://github.com/kubernetes/kubernetes/pull/99597/files#r631975169
> 
> It seems Adi is having some issues resolving this. Can we address it in a followup PR? This is not a user facing change, so it is safe to address later. I would like to get this PR merged asap because it is blocking multiple followup work.

@liggitt are you ok with this?

**Comment by adtac** [comment]
I think I've solved all issues now. PTAL.

tl;dr of what was blocking: API machinery seems to overwrite the TypeMeta.APIVersion field every time a conversion happens here:

https://github.com/kubernetes/kubernetes/blob/12e3860cb084ef99f79eb2bc7c123f9521342aff/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L544-L546

This makes sense in general, but we want something slightly different, so instead setting the APIVersion field in conversion.go (which doesn't work), we set it after we read the config file in post-conversion.

**Comment by ahg-g** [comment]
@liggitt I think Adi now addressed all the comments, please take a look.

**Review by msau42 — COMMENTED** [review]
latest api changes lgtm

File: cmd/kube-scheduler/app/options/configfile.go [hunk 21]
```diff
@@ -37,39 +40,61 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
...
+func encodeConfig(cfg *config.KubeSchedulerConfiguration) (*bytes.Buffer, error) {
+	buf := new(bytes.Buffer)
+	const mediaType = runtime.ContentTypeYAML
+	info, ok := runtime.SerializerInfoForMediaType(scheme.Codecs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return buf, fmt.Errorf("unable to locate encoder -- %q is not a supported media type", mediaType)
+	}
+
+	var encoder runtime.Encoder
+	switch cfg.TypeMeta.APIVersion {
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/configfile.go:70]: this is switching on encoding the internal config based on the external version it has in its apiVersion field? shouldn't we pick a default version to encode to (latest seems reasonable) like we did previously?

File: cmd/kube-scheduler/app/options/configfile.go [hunk 22]
```diff
@@ -37,39 +40,61 @@ func loadConfigFromFile(file string) (*kubeschedulerconfig.KubeSchedulerConfigur
...
 		return nil, err
 	}
-	if cfgObj, ok := obj.(*kubeschedulerconfig.KubeSchedulerConfiguration); ok {
+	if cfgObj, ok := obj.(*config.KubeSchedulerConfiguration); ok {
+		// We don't set this field in pkg/scheduler/apis/config/{version}/conversion.go
+		// because the field will be cleared later by API machinery during
+		// conversion. To work around that, we set the API version field of the
+		// internal object here to the version we read from file. We need this for
+		// later when we need to make decisions based on the external version.
+		cfgObj.TypeMeta.APIVersion = gvk.GroupVersion().String()
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/configfile.go:55]: do you have a pointer to the decisions made based on this field? it's pretty strange for the external version source to change the interpretation of the converted internal config

Comment (liggitt) [reply cmd/kube-scheduler/app/options/configfile.go:55]: if the only decision made is a pass/fail validation decision, that would be helpful to clarify here. we explicitly don't want the internal configuration getting interpreted differently based on version

File: pkg/scheduler/framework/runtime/framework.go [hunk 23]
```diff
@@ -415,13 +425,22 @@ func registerClusterEvents(name string, eventToPlugins map[framework.ClusterEven
...
 	}
 	// Use defaults from latest config API version.
-	gvk := v1beta1.SchemeGroupVersion.WithKind(name + "Args")
+	var gvk schema.GroupVersionKind
+	switch componentConfigVersion {
+	case v1beta1.SchemeGroupVersion.String():
+		gvk = v1beta1.SchemeGroupVersion.WithKind(name + "Args")
+	case v1beta2.SchemeGroupVersion.String():
+		gvk = v1beta2.SchemeGroupVersion.WithKind(name + "Args")
+	default:
```
Comment (liggitt) [inline pkg/scheduler/framework/runtime/framework.go:440]: should we default to this even if component config version is non-empty and is an unknown version?

File: pkg/scheduler/scheduler.go [hunk 24]
```diff
@@ -110,6 +111,14 @@ type schedulerOptions struct {
 // Option configures a Scheduler
 type Option func(*schedulerOptions)
 
+// WithComponentConfigVersion sets the component config version for the
+// KubeSchedulerConfiguration used.
```
Comment (liggitt) [inline pkg/scheduler/scheduler.go:115]: document examples to make it clear this is the full group/version

File: pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go [hunk 25]
```diff
@@ -70,6 +74,7 @@ func New(plArgs runtime.Object, handle framework.Handle) (framework.Plugin, erro
 	}
 	serviceLister := handle.SharedInformerFactory().Core().V1().Services().Lister()
 
+	klog.Warning("ServiceAffinity plugin is deprecated and will be removed in future versions")
```
Comment (liggitt) [inline pkg/scheduler/framework/plugins/serviceaffinity/service_affinity.go:77]: put the recommended replacement (InterPodAffinity) in the user-facing warning

File: pkg/scheduler/framework/plugins/nodepreferavoidpods/node_prefer_avoid_pods.go [hunk 26]
```diff
@@ -85,5 +89,6 @@ func (pl *NodePreferAvoidPods) ScoreExtensions() framework.ScoreExtensions {
 
 // New initializes a new plugin and returns it.
 func New(_ runtime.Object, h framework.Handle) (framework.Plugin, error) {
+	klog.Warning("NodePreferAvoidPods plugin is deprecated and will be removed in future versions")
```
Comment (liggitt) [inline pkg/scheduler/framework/plugins/nodepreferavoidpods/node_prefer_avoid_pods.go:92]: put the recommended replacement (node taints) in the user-facing warning

File: pkg/scheduler/framework/plugins/nodelabel/node_label.go [hunk 27]
```diff
@@ -46,6 +50,7 @@ func New(plArgs runtime.Object, handle framework.Handle) (framework.Plugin, erro
 		return nil, err
 	}
 
+	klog.Warning("NodeLabel plugin is deprecated and will be removed in future versions")
```
Comment (liggitt) [inline pkg/scheduler/framework/plugins/nodelabel/node_label.go:53]: mention the NodeAffinity plugin as a replacement

File: cmd/kube-scheduler/app/options/configfile.go (hunk 22 above)
Comment (ahg-g) [reply cmd/kube-scheduler/app/options/configfile.go:55]: There are two places, the validation and the plugin config defaulting logic inside the framework. The former will continue to exist, the latter will be removed as a followup to this PR (https://github.com/kubernetes/kubernetes/issues/102211).

File: pkg/scheduler/framework/runtime/framework.go (hunk 23 above)
Comment (ahg-g) [reply pkg/scheduler/framework/runtime/framework.go:440]: The original code was doing it regardless of `componentConfigVersion`, I think this is fine as is. In any case, this defaulting logic will be removed as a followup as part of https://github.com/kubernetes/kubernetes/issues/102211

File: cmd/kube-scheduler/app/options/configfile.go (hunk 21 above)
Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go:70]: since the internal object could've been converted from any version, there could be some information loss if we convert for example v1beta1 -> internal -> v1beta2.

to be quite honest, and I discussed this with @alculquicondor, I don't understand the purpose of this feature. who is using this and for what?

File: cmd/kube-scheduler/app/options/configfile.go (hunk 22 above)
Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go:55]: *three places. We also use this in `--write-config-to`. I've added a comment in the internal KubeSchedulerConfiguration with references to exactly what places use this field.

File: cmd/kube-scheduler/app/options/options.go [hunk 28]
```diff
@@ -133,14 +133,20 @@ func splitHostIntPort(s string) (string, int, error) {
...
 	kubeschedulerscheme.Scheme.Default(&versionedCfg)
 	cfg := kubeschedulerconfig.KubeSchedulerConfiguration{}
 	if err := kubeschedulerscheme.Scheme.Convert(&versionedCfg, &cfg, nil); err != nil {
 		return nil, err
 	}
+	// We don't set this field in pkg/scheduler/apis/config/{version}/conversion.go
+	// because the field will be cleared later by API machinery during
+	// conversion. To work around that, we set the API version field of the
+	// internal object here to the default version (latest). We need this for
+	// later when we need to make decisions based on the external version.
```
[the remark this answers is not in the document]
Comment (adtac) [reply cmd/kube-scheduler/app/options/options.go:148]: That is correct. I've added a comment in the internal KubeSchedulerConfiguration type definition.

File: cmd/kube-scheduler/app/options/configfile.go (hunk 21 above)
Comment (liggitt) [reply cmd/kube-scheduler/app/options/configfile.go:69]: > I don't understand the purpose of this feature. who is using this and for what?

isn't this the path taken when generating a config with `--write-config-to`?

Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go:69]: Yes.

Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go:69]: right, but I don't fully understand what that feature is used for. perhaps to generate a default config file that you can modify afterwards?

Comment (liggitt) [reply cmd/kube-scheduler/app/options/configfile.go:69]: > perhaps to generate a default config file that you can modify afterwards?

it's not just a default config file, it writes the effective configuration resulting from the flags you set, so someone transitioning from flags to a config file can add `--write-config-to=foo.yaml` to their existing flag-based invocation and get the equivalent config file.

Comment (alculquicondor) [reply cmd/kube-scheduler/app/options/configfile.go:69]: using latest version sgtm

Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go:69]: Done.

Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go]: actually, I'm having second thoughts about this: we necessarily need to switch when someone loads a v1beta1 config file that uses a deprecated plugin because the v1beta2 encoder cannot encode plugin args for deprecated plugins. am I missing something?

Comment (liggitt) [reply cmd/kube-scheduler/app/options/configfile.go]: I don't object to switching on APIVersion, my comment was limited to the behavior when APIVersion was empty (pick a default, instead of erroring)

Comment (adtac) [reply cmd/kube-scheduler/app/options/configfile.go]: ah I completely misread your comment. done.

**Comment by ahg-g** [comment]
/hold

pls squash :)
